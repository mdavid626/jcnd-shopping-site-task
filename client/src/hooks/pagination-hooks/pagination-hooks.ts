import { useCallback, useMemo } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';

export const usePagination = (): [number, () => void, () => void] => {
  const location = useLocation();
  const navigate = useNavigate();

  const currentPage = useMemo(() => {
    const query = new URLSearchParams(location.search);
    const currentPage = Number.parseInt(query.get('page')!, 10) || 0;
    return Math.max(currentPage, 0);
  }, [location.search]);

  const goNext = useCallback(() => {
    const query = new URLSearchParams(location.search);
    query.set('page', (currentPage + 1).toString());
    navigate({
      search: query.toString(),
    });
  }, [location.search, currentPage, navigate]);

  const goPrevious = useCallback(() => {
    const query = new URLSearchParams(location.search);
    query.set('page', (currentPage - 1).toString());
    navigate({
      search: query.toString(),
    });
  }, [location.search, currentPage, navigate]);

  return [currentPage, goNext, goPrevious];
};
