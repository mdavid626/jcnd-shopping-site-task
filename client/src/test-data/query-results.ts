export const productsQueryResult = {
  data: {
    products: {
      nodes: [
        {
          id: '63ea6485732ff5cf8b364431',
          priceInCents: 9699,
          name: 'Broccoli',
          inStock: 466,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364432',
          priceInCents: 3434,
          name: 'Broccolini',
          inStock: 223,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364433',
          priceInCents: 3692,
          name: 'Brussels Sprouts',
          inStock: 12,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364434',
          priceInCents: 8839,
          name: 'Cabbage',
          inStock: 200,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364435',
          priceInCents: 6828,
          name: 'Carrots',
          inStock: 151,
          __typename: 'Product',
        },
      ],
      pageInfo: { page: 1, numberOfPages: 15, __typename: 'PageInfo' },
      __typename: 'ProductsQueryResult',
    },
  },
};

export const placeOrderQueryResult1 = {
  data: { placeOrder: '63ede5e031f3fc4012a0951b' },
};
