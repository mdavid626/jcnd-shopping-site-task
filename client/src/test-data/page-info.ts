import { PageInfo } from '../types/page-info';

export const testPageInfo1: PageInfo = {
  page: 1,
  numberOfPages: 5,
};
