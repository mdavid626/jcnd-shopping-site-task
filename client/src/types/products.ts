import { PageInfo } from './page-info';

export type ProductCategory = 'VEGETABLES' | 'FRUITS' | 'CHEESE';

export type Product = {
  id: string;
  price: number;
  name: string;
  inStock: number;
};

export type ProductsQueryResult = {
  products: { nodes: Product[]; pageInfo: PageInfo };
};
