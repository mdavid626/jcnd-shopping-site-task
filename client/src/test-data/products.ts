import { Product, ProductsQueryResult } from '../types/products';

export const testProduct1: Product = {
  id: '63ea6485732ff5cf8b36442c',
  name: 'Artichoke',
  price: 46.36,
  inStock: 5,
};

export const testProduct2: Product = {
  id: '63ea6485732ff5cf8b36442d',
  name: 'Asparagus',
  price: 25.36,
  inStock: 3,
};

export const testProductQueryResult1: ProductsQueryResult = {
  products: {
    nodes: [testProduct1, testProduct2],
    pageInfo: {
      page: 1,
      numberOfPages: 2,
    },
  },
};
