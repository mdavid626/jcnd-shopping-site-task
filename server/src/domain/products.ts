import { ProductCategory } from '../types/product';
import sampleProducts from './sample-products';

const getProducts = (category: ProductCategory) => {
  return sampleProducts.filter((product) => product.category === category);
};

export default {
  getProducts,
};
