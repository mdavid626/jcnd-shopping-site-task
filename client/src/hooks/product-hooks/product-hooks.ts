import { ApolloError, gql, useQuery } from '@apollo/client';
import { ProductCategory, ProductsQueryResult } from '../../types/products';

export const GetProductsQuery = gql`
  query GetProductsQuery($category: ProductCategory!, $page: Int!) {
    products(category: $category, page: $page) {
      nodes {
        id
        price
        name
        inStock
      }
      pageInfo {
        page
        numberOfPages
      }
    }
  }
`;

export const useProducts = (
  category: ProductCategory,
  currentPage: number
): [ProductsQueryResult | undefined, ApolloError | undefined] => {
  const { data, error } = useQuery<ProductsQueryResult>(GetProductsQuery, {
    variables: {
      category,
      page: currentPage,
    },
    notifyOnNetworkStatusChange: true,
  });
  return [data, error];
};
