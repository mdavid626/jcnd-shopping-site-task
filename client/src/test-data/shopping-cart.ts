import { ShoppingCartItem } from '../types/shopping-cart';
import { testProduct1, testProduct2 } from './products';

export const testShoppingCartItem1: ShoppingCartItem = {
  product: testProduct1,
  amount: 3,
};

export const testShoppingCartItem2: ShoppingCartItem = {
  product: testProduct2,
  amount: 36,
};

export const testShoppingCart1: ShoppingCartItem[] = [
  testShoppingCartItem1,
  testShoppingCartItem2,
];
