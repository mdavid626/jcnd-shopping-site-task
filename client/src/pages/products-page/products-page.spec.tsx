import { cleanup, render, screen } from '@testing-library/react';
import React, { ReactNode } from 'react';
import Footer from '../../components/footer/footer';
import Header from '../../components/header/header';
import Pagination from '../../components/pagination/pagination';
import ProductItem from '../../components/product-item/product-item';
import Sidebar from '../../components/sidebar/sidebar';
import { usePagination } from '../../hooks/pagination-hooks/pagination-hooks';
import { useProducts } from '../../hooks/product-hooks/product-hooks';
import { testProductQueryResult1 } from '../../test-data/products';
import ProductsPage from './products-page';

jest.mock('../../components/footer/footer');
jest.mock('../../components/header/header');
jest.mock('../../components/pagination/pagination');
jest.mock('../../components/product-item/product-item');
jest.mock('../../components/sidebar/sidebar');
jest.mock('../../hooks/pagination-hooks/pagination-hooks');
jest.mock('../../hooks/product-hooks/product-hooks');

describe('products-page', () => {
  beforeEach(() => {
    (Footer as jest.Mock).mockReturnValue(<div>footer</div>);
    (Header as jest.Mock).mockReturnValue(<div>header</div>);
    (Pagination as jest.Mock).mockReturnValue(<div>pagination</div>);
    (ProductItem as jest.Mock).mockReturnValue(<div>product-item</div>);
    (Sidebar as jest.Mock).mockImplementation(
      ({ children }: { children: ReactNode }) => <div>sidebar {children}</div>
    );
    (usePagination as jest.Mock).mockReturnValue([1, jest.fn(), jest.fn()]);
    (useProducts as jest.Mock).mockReturnValue([
      testProductQueryResult1,
      undefined,
    ]);
  });
  afterEach(cleanup);
  afterEach(jest.resetAllMocks);

  it('should render', () => {
    const goNext = jest.fn();
    const goPrevious = jest.fn();
    (usePagination as jest.Mock).mockReturnValue([1, goNext, goPrevious]);
    const { asFragment } = render(<ProductsPage category="VEGETABLES" />);
    expect(asFragment()).toMatchSnapshot();
    expect(usePagination).toHaveBeenCalledWith();
    expect(useProducts).toHaveBeenCalledWith('VEGETABLES', 1);
    expect(Header).toHaveBeenCalledWith({}, {});
    expect(Footer).toHaveBeenCalledWith({}, {});
    expect(Pagination).toHaveBeenCalledWith(
      {
        pageInfo: testProductQueryResult1.products.pageInfo,
        goNext,
        goPrevious,
        className: 'ProductsPage-pagination',
      },
      {}
    );
    expect(ProductItem).toHaveBeenCalledTimes(2);
    expect(ProductItem).toHaveBeenCalledWith(
      { product: testProductQueryResult1.products.nodes[0] },
      {}
    );
    expect(ProductItem).toHaveBeenCalledWith(
      { product: testProductQueryResult1.products.nodes[1] },
      {}
    );
    expect(Sidebar).toHaveBeenCalledWith(
      { className: 'ProductsPage-sidebar', children: expect.any(Object) },
      {}
    );
  });

  it('should show spinner when loading', () => {
    (useProducts as jest.Mock).mockReturnValue([undefined, undefined]);
    const { asFragment } = render(<ProductsPage category="VEGETABLES" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should show error when error', () => {
    (useProducts as jest.Mock).mockReturnValue([
      undefined,
      new Error('test error'),
    ]);
    render(<ProductsPage category="VEGETABLES" />);
    expect(screen.getByText('test error')).toBeVisible();
  });
});
