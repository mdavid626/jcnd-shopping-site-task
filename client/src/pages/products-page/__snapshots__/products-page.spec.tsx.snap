// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`products-page should render 1`] = `
<DocumentFragment>
  <div
    class="ProductsPage"
  >
    <div>
      header
    </div>
    <div>
      sidebar 
      <div
        class="ProductsPage-content"
      >
        <div>
          pagination
        </div>
        <div
          class="ProductsPage-products"
        >
          <div>
            product-item
          </div>
          <div>
            product-item
          </div>
        </div>
      </div>
    </div>
    <div>
      footer
    </div>
  </div>
</DocumentFragment>
`;

exports[`products-page should show spinner when loading 1`] = `
<DocumentFragment>
  <div
    class="ProductsPage"
  >
    <div>
      header
    </div>
    <div>
      sidebar 
      <div
        class="ProductsPage-content"
      >
        <div>
          pagination
        </div>
        <img
          alt="loading spinner"
          class="PageLoader-spinner"
          data-testid="PageLoader-spinner"
          src="spinner.svg"
        />
      </div>
    </div>
    <div>
      footer
    </div>
  </div>
</DocumentFragment>
`;
